src/placement/mapless_jump_map.c

```diff
@@ -0,0 +1,1404 @@
+ *
+ * \param[in]   root            This is the root of the pool map tree structure
+ *                              containing all the fault domain information.
+ * \param[out]  layout          This structure will contained the ordered list
+ *                              of targets to be used for the object layout.
+ * \param[in]   com_map         This is the collision map used during the
+ *                              placement process for bookkeeping and
+ *                              determining which domains have been used.
+ * \param[in]   grp_cnt       This number of redundancy groups for this
+ *                              objects layout.
+ * \param[in]   grp_size      The size of each redundancy group in the layout.
+ *                              This will be multiplied with group_cnt to get
+ *                              the total number of targets required for
+ *                              placement.
+ * \param[in]   dom_map_size    The number of non-leaf nodes in the pool map.
+ *                              This value will be used to initialize the array
+ *                              that holds the statuses of whether the node has
+ *                              been used or not.
+ * \param[in]   cnt_map_size    The number of child domains used for a given
+ *                              non-leaf node in the pool map. When this equals
+ *                              the total number of child nodes we reset the
+ *                              status showing the nodes have been used.
+ * \param[out]  rebuild_list      The list of nodes that needs to be rebuilt
+ *                              during the rebuild process. This is only
+ *                              returned when this function is called from the
+ *                              rebuild function and will otherwise be NULL.
+ *
+ * \return                      Returns the number of targets that need to be
+ *                              rebuilt. This number will be ignored when called
+ *                              from functions other than find rebuild.
+ */
+static int
+get_object_layout(struct pl_mapless_map *mmap, struct pl_obj_layout *layout,
+		struct mapless_obj_placement *mop, daos_obj_id_t oid,
+		d_list_t *remap_list, struct daos_obj_md *md)
+{
+	struct pool_target	*target;
+	uint8_t			*dom_used;
+	uint32_t		*used_targets;
+	uint64_t		key;
+	struct pool_domain 	*root;
+	int i, j, k, rc;
+
+	/* Set the pool map version */
+	layout->ol_ver = pl_map_version(&(mmap->mmp_map));
+
+	j = 0;
+	k = 0;
+	key = oid.lo;
+
+	rc = 0;
+	target = NULL;
+
+	D_ALLOC_ARRAY(dom_used, mmap->dom_used_length);
+	D_ALLOC_ARRAY(used_targets, layout->ol_nr + 1);
+
+	if (dom_used == NULL)
+		return -DER_NOMEM;
+	if (used_targets == NULL)
+		return -DER_NOMEM;
+
+	if (daos_obj_id2class(oid) == DAOS_OC_R3S_SPEC_RANK ||
+			daos_obj_id2class(oid) == DAOS_OC_R1S_SPEC_RANK ||
+		    	daos_obj_id2class(oid) == DAOS_OC_R2S_SPEC_RANK) {
```

Comment:
(style)  code indent should use tabs where possible
(style)  please, no space before tabs